crates/edr_solidity/src/vm_tracer.rs

```diff
@@ -0,0 +1,276 @@
+        for msg in &trace.messages {
+            match msg.clone() {
+                edr_evm::trace::TraceMessage::Before(before) => {
+                    self.add_before_message(before);
+                }
+                edr_evm::trace::TraceMessage::Step(step) => {
+                    self.add_step(step);
+                }
+                edr_evm::trace::TraceMessage::After(after) => {
+                    self.add_after_message(after.execution_result);
+                }
+            }
+        }
+    }
+
+    fn should_keep_tracing(&self) -> bool {
+        self.last_error.is_none()
+    }
+
+    fn add_before_message(&mut self, message: BeforeMessage) {
+        if !self.should_keep_tracing() {
+            return;
+        }
+
+        let trace: MessageTrace;
+
+        if message.depth == 0 {
+            self.message_traces.clear();
+            self.tracing_steps.clear();
+        }
+
+        if message.to.is_none() {
+            let create_trace = CreateMessageTrace {
+                base: BaseEvmMessageTrace {
+                    base: BaseMessageTrace {
+                        depth: message.depth,
+                        value: message.value,
+                        exit: ExitCode::Success,
+                        return_data: Bytes::new(),
+                        gas_used: 0,
+                    },
+                    code: message.data,
+                    steps: vec![],
+                    number_of_subtraces: 0,
+                    // this was not in the original code - assumed to be None/undefined?
+                    bytecode: None,
+                },
+
+                deployed_contract: None,
+            };
+
+            trace = MessageTrace::Create(create_trace);
+        } else {
+            // TODO: Make this nicer and make sure the U160 logic/precompile comparison
+            // logic works
+            let to = message.to.unwrap();
+            let to_as_bigint = U160::from_be_bytes(**to);
+
+            if to_as_bigint <= U160::from(self.max_precompile_number) {
+                let precompile_trace = PrecompileMessageTrace {
+                    base: BaseMessageTrace {
+                        value: message.value,
+                        exit: ExitCode::Success,
+                        return_data: Bytes::new(),
+                        depth: message.depth,
+                        gas_used: 0,
+                    },
+                    // The max precompile number is 10, so we can safely unwrap here
+                    precompile: to_as_bigint.as_limbs()[0] as u32,
+                    calldata: message.data,
+                };
+
+                trace = MessageTrace::Precompile(precompile_trace);
+            } else {
+                // if we enter here, then `to` is not None, therefore
+                // `code_address` and `code` should be Some
+                // TODO (if not true, then we need to return and set the error message)
```

Comment:
Todo. Are you planning to resolve this before merging?

If not, could you please create a GitHub issue for this and link to its URL here?

Reply:
Thanks, this was a leftover TODO; I already did that part below.